### ISSUE TITLE
Unable to Access Application 

### ISSUE THREAD

**Issue by rambabu0063** [mspnp/aks-secure-baseline#20]
Hi Team,

I have deployed cluster with workload successfully but I am unable to access application running on cluster.
I have followed the below link:
https://github.com/mspnp/aks-secure-baseline/blob/main/10-validation.md

please find the Attached screenshot.
<img width="652" alt="sight-access" src="https://user-images.githubusercontent.com/57993345/88547373-422b4000-d03b-11ea-90df-b93fd024fb89.PNG">
<img width="446" alt="Host-file" src="https://user-images.githubusercontent.com/57993345/88547387-49524e00-d03b-11ea-8697-fb6c80bf6e0a.PNG">

**Comment by ferantivero** [comment]
thanks for sharing this @rambabu0063, probably it is backend which is currently in an un-healthy state. Therefore, please let us share couple of troubleshooting steps to find the possible root cause:

1. go to your Azure Application Gateway instance to check whether it is healthy state: `Azure Application Gateway`-> `Monitoring` ->  `Backend Health`
1. explore the CSI logs to ensure your TLS certs are obtained from Azure Key Vault: `kubectl logs <csi-secrets-store-name> -n cluster-baseline-settings -c secrets-store`
1. explore the traefik logs: `kubeclt logs <traefik-pod-name> -n a0008`

**Comment by rambabu0063** [comment]
Hi @ferantivero 

thanks for your response.

I have checked and found that Application Gateway is unhealthy.
<img width="955" alt="APG-status" src="https://user-images.githubusercontent.com/57993345/88642956-bbc83a00-d0de-11ea-863b-54009cc7e1b8.PNG">

**Comment by ferantivero** [comment]
thanks for sharing your progress @rambabu0063, please let's continue gathering info by executing steps 2 & 3.

**Comment by rambabu0063** [comment]
Hi,

I have re-deployed the solution and i  did not get output for the resource. it was throwing message like "no matching resource found".

I am following the link:  https://github.com/mspnp/aks-secure-baseline/blob/main/08-secret-managment-and-ingress-controller.md

I am getting this issue at step 6:
I have waited for 50 min . I was getting same output after 50 min also.

<img width="944" alt="no-matching-resource-found" src="https://user-images.githubusercontent.com/57993345/88818799-bc46fa80-d1dc-11ea-88ac-a1b595262ba4.PNG">
<img width="469" alt="no-resources-found-in-namespaces" src="https://user-images.githubusercontent.com/57993345/88818832-c8cb5300-d1dc-11ea-8947-ac330ce1b57e.PNG">

**Comment by ferantivero** [comment]
thanks for sharing this @rambabu0063, the command you've executed returns a result if `traefik` is up & running after a certain period of time (`90 s` in this case). That said, it is telling us that for some reason `traefik` couldn't reach the `ready` state and that's why it hangs in there. 

Could you please execute the following commands to try to troubleshoot this in your environment?

1. explore the CSI logs to ensure your TLS certs are obtained from Azure Key Vault: `kubectl logs <csi-secrets-store-name> -n cluster-baseline-settings -c secrets-store`
1. inspect the SecretProvideClass object and see the Azure KeyVault Name and Tenant Id are properly set: `kubectl get SecretProviderClass aks-ingress-contoso-com-tls-secret-csi-akv -n a0008`
1. explore the traefik logs: `kubeclt logs <traefik-pod-name> -n a0008`

**Comment by rambabu0063** [comment]
Hi ,
 
<img width="777" alt="cs-secret-logs" src="https://user-images.githubusercontent.com/57993345/88926916-48682900-d294-11ea-98b2-9dc9b67cf5e0.PNG">
<img width="465" alt="no-resource-found" src="https://user-images.githubusercontent.com/57993345/88926920-4a31ec80-d294-11ea-84bf-f03ebde27821.PNG">
<img width="950" alt="SecretProvideClassobject" src="https://user-images.githubusercontent.com/57993345/88926921-4aca8300-d294-11ea-9c5c-651a73e420fb.PNG">



1. I have exported the logs and attached an Image
2. I can see the same keyvault name and tenant id has configured
3. there is no pods created under namespaces a0008 , unable to fetch log.

**Comment by ferantivero** [comment]
thanks for sharing this @rambabu0063,

could you please check if you have any deploy in `a0008` namespace? if not, please let's inspect for any violation that azure-policy/Gatekeeper could've detected and as consequence prevented `traefik` from getting deployed:

1. `kubectl get deploy -n a0008`
1. `kubectl logs <azure-policy-pod-name> -n kube-system`
1. `kubectl logs <gatekeeper-pod-name> -n gatekeeper-system`

**Comment by rambabu0063** [comment]
Hi @ferantivero,
thanks for you quick reply.

there is  no resource created under namespace a0008.

please find the attached requested logs.

[Azure-policy-log.txt](https://github.com/mspnp/aks-secure-baseline/files/5001557/Azure-policy-log.txt)
[gatekeeper-pod-logs.txt](https://github.com/mspnp/aks-secure-baseline/files/5001559/gatekeeper-pod-logs.txt)
<img width="442" alt="kubectl get deploy -n a0008" src="https://user-images.githubusercontent.com/57993345/88930331-faa1ef80-d298-11ea-96b3-63cdaebba7ef.PNG">
<img width="563" alt="no-resources-found-ns-a0008" src="https://user-images.githubusercontent.com/57993345/88930352-02fa2a80-d299-11ea-8d61-a75e199d4ac8.PNG">

**Comment by ferantivero** [comment]
@rambabu0063 please let’s try to inspect the deploy to pinpoint why traefik are not getting scheduled:

`kubectl describe deploy traefik-ingress-controller`

**Comment by rambabu0063** [comment]
Hi @ferantivero ,

please find the attached logs.
[traefik-ingress-controller-logs.txt](https://github.com/mspnp/aks-secure-baseline/files/5001597/traefik-ingress-controller-logs.txt)

**Comment by ferantivero** [comment]
@rambabu0063 everything is still looking good, which is a favorable sign. Therefore, please let's keep digging a bit more by executing the following 3 commands:

1. `kubectl logs <azure-policy-pod-name> -n kube-system`
1. `kubectl logs <gatekeeper-pod-name> -n gatekeeper-system`
1. `kubectl get events --all-namespaces`

**Comment by rambabu0063** [comment]
Hi @ferantivero ,

please find the node pool configuration details.
[node-pool configuration logs.txt](https://github.com/mspnp/aks-secure-baseline/files/5001782/node-pool.configuration.logs.txt)

**Comment by ferantivero** [comment]
@rambabu0063, I've just noticed you had already shared azure-policy and gatekeeper logs. So please let's take look at the last command only:

`kubectl get events --all-namespaces`

**Comment by rambabu0063** [comment]
Hi  @ferantivero ,

please find the event logs.
<img width="946" alt="event-logs" src="https://user-images.githubusercontent.com/57993345/88936010-faf1b900-d29f-11ea-9653-54caec6bbc50.PNG">

**Comment by ferantivero** [comment]
thanks @rambabu0063 you throw light on the issue here 😸. It seems that `Gatekeeper` is preventing `traefik` from getting scheduled since the memory limit is higher than the allowed, which is s bit confusing as it doesn't seem to be the actual case based on the log entry message.

As a conclusion til this point, this is why `traefik` is not running in the cluster and as a consequence having Azure Application Gateway with un-healthy state.

Since this is a preview feature, you may want to disable this add-on temporary, and proceed without `Azure Policy for Kubernetes` for now or wait for us to investigate this issue. In case you decide to disable this:

1. `az aks disable-addons --addons azure-policy --name <aks-cluster-name> --resource-group <resource-group>`
1. re-deploy `traefik` - it is more like an _optional_ step as this may not be required

**Comment by rambabu0063** [comment]
Hi @ferantivero,

facing issue while executing the given command
<img width="886" alt="azure-policy-disable-error" src="https://user-images.githubusercontent.com/57993345/88942846-6c356a00-d2a8-11ea-95bc-46215c6a76e3.PNG">

**Comment by ferantivero** [comment]
@rambabu0063 sorry to hear you faced that issue. Please let's try to follow the instructions on [how to do this from Azure Portal](https://docs.microsoft.com/en-us/azure/governance/policy/concepts/policy-for-kubernetes#remove-the-add-on-from-aks).

**Comment by rambabu0063** [comment]
Hi @ferantivero ,

thanks for your support.

I have disable the Azure policy and re-deployed the traefik. I am able to access the Application now.

<img width="703" alt="App-UI" src="https://user-images.githubusercontent.com/57993345/89060937-7e370b80-d381-11ea-9b30-80341ab9010c.PNG">

could you please suggest changes required  if we want to proceed with Azure policy.

**Comment by ckittel** [comment]
That's great news @rambabu0063.  We've got a task internally to work through this part of the reference implementation so other customers don't run into the same issues.  Your feedback along the way has been extremely valuable.  We'll likely be working on that starting next week.  I want to make sure we keep this issue OPEN so that if other customers are seeing the same, they'll notice the open issue and be able to read through it.

**Comment by ckittel** [comment]
@hallihan as recently made a change to the Azure Policy defination to fix the problem you (and others) were running into.  Policy changes take about 30 minutes to apply to your cluster.  Our Azure Policy in AKS documentation currently says 15, but in our testing, it's been more like 30.  This should unblock all part of this issue, and as such I'm going to close the issue.  Any new things that pop up, please go ahead and open a new issue.  Cheers.
